File: 01/Part2.ts
import * as fs from 'fs'

function sumMatch(str: string) {
  let list = str.split('').map( s => Number.parseInt(s))
  let sum = 0
  let length = list.length, dist = list.length/2
  for (let index in list) {
    let place = Number.parseInt(index)
      , pos = ((place + dist) % length)
    pos = pos >= 0 ? pos : length - 1
    let digit = list[place]
    let other = list[pos]
    sum += digit === other ? digit : 0
  }
  return sum
}

console.log(sumMatch(fs.readFileSync('list.txt','utf8').replace(/[^0-9]/g,'')))

File: 01/Part2.lua
function sumMatch(str)
  local list, sum = {}, 0
  for char in str:gmatch('%d') do
    list[#list+1] = tonumber(char)
  end
  local length, dist = #list, #list/2
  for i, digit in ipairs(list) do
    local pos = ((i+dist) % length)
    pos = pos == 0 and length or pos
    local other = list[pos]
    if other then
      sum = digit == other and (sum+digit) or (sum)
    else
      sum = digit == list[1] and (sum+digit) or (sum)
    end
  end

  return sum
end

-- print(sumMatch '1212')
-- print(sumMatch '1221')
-- print(sumMatch '123425')
-- print(sumMatch '123123')
-- print(sumMatch '12131415')
local file = io.open('list.txt')
print(sumMatch(file:read('*all'):gsub('[^%d]','')))
file:close()